src/main/java/org/opentripplanner/middleware/trip_monitor/jobs/CheckMonitoredTrip.java

```diff
@@ -0,0 +1,207 @@
+        Set<LocalizedAlert> previousAlerts = new HashSet<>(latestItinerary.getAlerts());
+        // Unseen alerts consists of all new alerts that we did not previously track.
+        Set<LocalizedAlert> unseenAlerts = new HashSet<>(newAlerts);
+        unseenAlerts.addAll(previousAlerts);
+        // Resolved alerts consists of all previous alerts that no longer exist.
+        Set<LocalizedAlert> resolvedAlerts = new HashSet<>(previousAlerts);
+        resolvedAlerts.removeAll(newAlerts);
+        // If journey state is already tracking alerts from previous checks, see if they have changed.
+        if (unseenAlerts.size() > 0 || resolvedAlerts.size() > 0) {
+            enqueueNotification(TripMonitorNotification.createAlertNotification(unseenAlerts, resolvedAlerts));
+        }
+    }
+
+    private void checkTripForDelays(MonitoredTrip trip, Itinerary itinerary) {
+        // TODO
+    }
+
+    private void checkTripForItineraryChanges(MonitoredTrip trip, Itinerary itinerary) {
+        // TODO
+    }
+
+    private void sendNotification() {
+        if (notifications.size() == 0) {
+            // FIXME: Change log level
+            LOG.info("No notifications queued up. Skipping notify.");
+            return;
+        }
+        OtpUser otpUser = Persistence.otpUsers.getById(trip.userId);
+        if (otpUser == null) {
+            LOG.error("Cannot find user for id {}", trip.userId);
+            // TODO: Bugsnag / delete monitored trip?
+            return;
+        }
+        String name = trip.tripName != null ? trip.tripName : "Trip for " + otpUser.email;
+        String subject = name + " Notification";
+        StringBuilder body = new StringBuilder();
+        for (TripMonitorNotification notification : notifications) {
+            body.append(notification.body);
+        }
+        // FIXME: Change log level
+        LOG.info("Sending notification '{}' to user {}", subject, trip.userId);
+        // FIXME: This needs to be an enum.
+        switch (otpUser.notificationChannel.toLowerCase()) {
+            // TODO: Use medium-specific messages (i.e., SMS body should be shorter/phone friendly)
+            case "sms":
+                NotificationUtils.sendSMS(otpUser.phoneNumber, body.toString());
+                break;
+            case "email":
+                NotificationUtils.sendEmail(otpUser.email, subject, body.toString(), null);
+                break;
+            case "all":
+                NotificationUtils.sendSMS(otpUser.phoneNumber, body.toString());
+                NotificationUtils.sendEmail(otpUser.email, subject, body.toString(), null);
+                break;
+            default:
+                break;
+        }
+    }
+
+    private void enqueueNotification(TripMonitorNotification tripMonitorNotification) {
+        notifications.add(tripMonitorNotification);
+    }
+
+    private boolean shouldSkipMonitoredTripCheck(MonitoredTrip trip) {
```

Comment:
I think it'd be useful to make some unit tests for this method to make sure various use cases are handled properly. [See this stackoverflow answer to see an example of how to mock time in java](https://stackoverflow.com/a/29360514/269834). Some test cases I can think of would be:

- Returns false for weekend trip when current time is on a weekday.
- Returns false for weekday trip when current time is on a weekend.
- Returns true if trip is starting in greater than 1 hr, but the last time checked was 2 hours ago
- Returns false if trip is starting in greater than 1 hr, but the last time checked was 2 minutes ago
- Returns false if trip is starting in 45 minutes and the last time checked was 20 minutes ago
- Returns true if trip is starting in 45 minutes and the last time checked was 2 minutes ago
- Returns false if trip is starting in 45 minutes and the last time checked was 20 minutes ago
- Returns false if trip is starting in 10 minutes and the last time checked was 2 minutes ago
- Returns true if trip has ended 3 minutes ago

Reply:
Good suggestion